### PR TITLE
Detect and handle various environments' global.

### LINKED ISSUE #43: [BUG] ember-cli-shims 0.0.6 and above is not compatible with Node
Using ember-cli-shims version of 0.0.6 and above in Fastboot (Node) environments gives an error:
`self is not defined`

This prevents the vendor file to be loaded in the sandbox. The error seems to be coming https://github.com/ember-cli/ember-cli-shims/blob/master/app-shims.js#L247 where it is using self.jQuery. While this would work in browsers, self is not defined in Node environments. 

We should check for existence of `self` before assigning the jQuery module.

I propose something like:

``` javascript
var jQueryVersion;
if (self) {
  jQueryVersion = self.jQuery;
} else {
 jQueryVersion = jQuery;
}
generateModule('jquery', { 'default': jQueryVersion });
```

### PR DESCRIPTION
**Pull request by rwjblue**
Kindly lifted from Ember (`packages/ember-environment/lib/global.js`) which lifted it from Lodash.
- Detects correct global in most environments (Node, browser, service worker, etc).
- Anchors all of the global references (i.e. `Ember`, `jQuery`) to the detected global.

Closes #43
Closes #44

/cc @kratiahuja @krisselden @chadhietala @stefanpenner

### CODE REVIEW

File: app-shims.js [hunk 1]
```diff
@@ -1,207 +1,227 @@
...
+      // element ids can ruin global miss checks
+      function checkElementIdShadowing(value) {
+        return (value && value.nodeType === undefined) ? value : undefined;
+      }
+
+      // export real global
+      return checkGlobal(checkElementIdShadowing(typeof global === 'object' && global)) ||
+             checkGlobal(typeof self === 'object' && self) ||
+             checkGlobal(typeof window === 'object' && window) ||
+             new Function('return this')(); // eval outside of strict mode
```
Comment (chadhietala) [inline app-shims.js]: CSP is going to have opinions about this, but I'm in general 👍

Comment (rwjblue) [inline app-shims.js]: Ya, I can remove it (it is included in Ember by default also), but if we do not have `self` (workers) or `window` (normal browser) how can we need to deal with CSP?

Comment (stefanpenner) [inline app-shims.js]: @rwjblue worker has `self`

Comment (stefanpenner) [inline app-shims.js]: I think CSP is runtime not static, and in if someone has CSP enabled but no window or no self, then im not sure if we can help them...

**Comment by stefanpenner** [comment]
> Anchors all of the global references (i.e. Ember, jQuery) to the detected global.

Why is this important? Over just lexical scoping?

**Comment by Turbo87** [comment]
closing due to inactivity

**Pull request closed without merging**